# Objetivos do estudo do teste A/B

## Objetivos do Teste A/B

**Objetivo principal:** Testar mudanças relacionadas à introdução de um sistema de recomendação melhorado

**Métricas avaliadas:** Usuários mostram uma conversão melhor nas visualizações de página do produto (o evento product_page event) e de compras (purchase).

**Critério de sucesso:** Em cada etapa do funil login → product_page → purchase, haverá ao menos 10% de aumento.

**Período de avaliação:** Em até 14 dias após o cadastro

**Público-alvo:** 
- ***Quem são os usuários?*** 
São novos usuários da região da UE, cadastrados entre 07/12 e 21/12/2020.

- ***Como foram divididos?*** 
Foram divididos em grupo A (controle) e grupo B (teste)]

# Bibliotecas e Funções usadas

## Bibliotecas:

In [1]:
import pandas as pd
import plotly.express as px
from statsmodels.stats.proportion import proportions_ztest

## Funções:

#### Função ***teste_z***: Essa função executa o teste z para a igualdade de proporções

In [2]:
def teste_z(grupo_1_evento: int, grupo_2_evento:int , total_grupo_1: int, total_grupo_2: int, alpha=0.01):
    '''
    Essa função executa o teste z para a igualdade de proporções

    INPUTS:
    grupo_1_evento: Quantidade de Usuários do grupo 1 que realizaram o evento (int)
    grupo_2_evento: Quantidade de Usuários do grupo 2 que realizaram o evento (int)
    total_grupo_1: Total de usuários do grupo 1 (int)
    total_grupo_2: Total de usuários do grupo 2 (int)
    '''
    
    # Teste Z
    z_stat, p_value = proportions_ztest([grupo_1_evento, grupo_2_evento], [total_grupo_1, total_grupo_2])
    print(f"Z-statistic: {z_stat:.4f}")
    print(f"P-value: {p_value:.4f}")
    print(f"Alpha: {alpha}")
    
    if p_value < alpha:
        return  '❌Rejeitamos a hipótese nula: A conversão entre os grupos foi DIFERENTE'
    else:
        return  '✅Aceitamos a hipótese nula: A conversão entre os grupos foi IGUAL, logo o teste não trouxe melhorias.'

# 1. Importando e tratando os dados:

## 1.1. Otimizando e Importando os dados

### 1.1.1. Otimizando os dados para importa-los:

- ***Importando uma amostra dos dados para otimiza-los***

In [3]:
inicio = '../datasets/'

marketing_amostra = pd.read_csv(inicio + 'ab_project_marketing_events_us.csv', nrows = 100)

events_amostra = pd.read_csv(inicio + 'final_ab_events_upd_us.csv', nrows = 100)

users_amostra = pd.read_csv(inicio + 'final_ab_new_users_upd_us.csv', nrows = 100)

participants_amostra = pd.read_csv(inicio + 'final_ab_participants_upd_us.csv', nrows = 100)

- ***Imprimindo as informações gerais e as primeiras linhas das amostras:***

In [4]:
print('_____________________________________Dataframe marketing_amostra:____________________________________________ \n')
print('- As colunas start_dt e finish_dt devem ser do tipo datetime; \n- A coluna regions deve ser do tipo categorica;\n')
marketing_amostra.info(memory_usage='deep')
marketing_amostra.head()

_____________________________________Dataframe marketing_amostra:____________________________________________ 

- As colunas start_dt e finish_dt devem ser do tipo datetime; 
- A coluna regions deve ser do tipo categorica;

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14 entries, 0 to 13
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   name       14 non-null     object
 1   regions    14 non-null     object
 2   start_dt   14 non-null     object
 3   finish_dt  14 non-null     object
dtypes: object(4)
memory usage: 3.6 KB


,name,regions,start_dt,finish_dt
0,Christmas&New Year Promo,"EU, N.America",2020-12-25,2021-01-03
1,St. Valentine's Day Giveaway,"EU, CIS, APAC, N.America",2020-02-14,2020-02-16
2,St. Patric's Day Promo,"EU, N.America",2020-03-17,2020-03-19
3,Easter Promo,"EU, CIS, APAC, N.America",2020-04-12,2020-04-19
4,4th of July Promo,N.America,2020-07-04,2020-07-11


In [5]:
print('_______________________________________ Dataframe events_amostra: ______________________________________________\n')
print('- A coluna event_dt deve ser do tipo datetime; \n- A coluna event_name deve ser do tipo categorica;\n')
events_amostra.info(memory_usage='deep')
events_amostra.head()


_______________________________________ Dataframe events_amostra: ______________________________________________

- A coluna event_dt deve ser do tipo datetime; 
- A coluna event_name deve ser do tipo categorica;

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100 entries, 0 to 99
Data columns (total 4 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   user_id     100 non-null    object 
 1   event_dt    100 non-null    object 
 2   event_name  100 non-null    object 
 3   details     100 non-null    float64
dtypes: float64(1), object(3)
memory usage: 19.5 KB


,user_id,event_dt,event_name,details
0,E1BDDCE0DAFA2679,2020-12-07 20:22:03,purchase,99.99
1,7B6452F081F49504,2020-12-07 09:22:53,purchase,9.99
2,9CD9F34546DF254C,2020-12-07 12:59:29,purchase,4.99
3,96F27A054B191457,2020-12-07 04:02:40,purchase,4.99
4,1FD7660FDF94CA1F,2020-12-07 10:15:09,purchase,4.99


In [6]:
print('______________________________________ Dataframe users_amostra: _________________________________________\n')
print('- A coluna first_date deve ser do tipo datetime; \n- As colunas region e device devem ser do tipo categorica;\n')
users_amostra.info(memory_usage='deep')
users_amostra.head()


______________________________________ Dataframe users_amostra: _________________________________________

- A coluna first_date deve ser do tipo datetime; 
- As colunas region e device devem ser do tipo categorica;

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100 entries, 0 to 99
Data columns (total 4 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   user_id     100 non-null    object
 1   first_date  100 non-null    object
 2   region      100 non-null    object
 3   device      100 non-null    object
dtypes: object(4)
memory usage: 22.6 KB


,user_id,first_date,region,device
0,D72A72121175D8BE,2020-12-07,EU,PC
1,F1C668619DFE6E65,2020-12-07,N.America,Android
2,2E1BF1D4C37EA01F,2020-12-07,EU,PC
3,50734A22C0C63768,2020-12-07,EU,iPhone
4,E1BDDCE0DAFA2679,2020-12-07,N.America,iPhone


In [7]:
print('_________________________________________ Dataframe participants_amostra: ____________________________________\n')
print('- As colunas group e ab_test devem ser do tipo categorica;\n')
participants_amostra.info(memory_usage='deep')
participants_amostra.head()

_________________________________________ Dataframe participants_amostra: ____________________________________

- As colunas group e ab_test devem ser do tipo categorica;

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100 entries, 0 to 99
Data columns (total 3 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   user_id  100 non-null    object
 1   group    100 non-null    object
 2   ab_test  100 non-null    object
dtypes: object(3)
memory usage: 18.4 KB


,user_id,group,ab_test
0,D1ABA3E2887B6A73,A,recommender_system_test
1,A7A3664BD6242119,A,recommender_system_test
2,DABC14FDDFADD29E,A,recommender_system_test
3,04988C5DF189632E,A,recommender_system_test
4,4FF2998A348C484F,A,recommender_system_test


- ***Otimizando os dados das amostras:***

In [8]:
# Marketing:
marketing_amostra['start_dt'] = pd.to_datetime(marketing_amostra['start_dt'], format='%Y-%m-%d')
marketing_amostra['finish_dt'] = pd.to_datetime(marketing_amostra['finish_dt'], format='%Y-%m-%d')

marketing_amostra['regions'] = marketing_amostra['regions'].astype('category')

# Events:
events_amostra['event_dt'] = pd.to_datetime(events_amostra['event_dt'], format='%Y-%m-%d %H:%M:%S')

events_amostra['event_name'] = events_amostra['event_name'].astype('category')

# Users:
users_amostra['first_date'] = pd.to_datetime(users_amostra['first_date'], format='%Y-%m-%d')

users_amostra['region'] = users_amostra['region'].astype('category')
users_amostra['device'] = users_amostra['device'].astype('category')

# Participants:
participants_amostra['group'] = participants_amostra['group'].astype('category')
participants_amostra['ab_test'] = participants_amostra['ab_test'].astype('category')


In [9]:
print(marketing_amostra.info(memory_usage='deep'))
print(events_amostra.info(memory_usage='deep'))
print(users_amostra.info(memory_usage='deep'))
print(participants_amostra.info(memory_usage='deep'))

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14 entries, 0 to 13
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype         
---  ------     --------------  -----         
 0   name       14 non-null     object        
 1   regions    14 non-null     category      
 2   start_dt   14 non-null     datetime64[ns]
 3   finish_dt  14 non-null     datetime64[ns]
dtypes: category(1), datetime64[ns](2), object(1)
memory usage: 1.9 KB
None
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100 entries, 0 to 99
Data columns (total 4 columns):
 #   Column      Non-Null Count  Dtype         
---  ------      --------------  -----         
 0   user_id     100 non-null    object        
 1   event_dt    100 non-null    datetime64[ns]
 2   event_name  100 non-null    category      
 3   details     100 non-null    float64       
dtypes: category(1), datetime64[ns](1), float64(1), object(1)
memory usage: 8.3 KB
None
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100 entries

Conclusão: 

Conseguimos reduzir o uso de memória dos nossos dados em quase 3 vezes, o que é excelente.

### 1.1.2. Importando os dados completos (aplicando a nossa otimização):

In [10]:
# Carregando os dados:

inicio = '../datasets/'

marketing = pd.read_csv(inicio + 'ab_project_marketing_events_us.csv', dtype = {'regions':'category'}, parse_dates=['start_dt','finish_dt'])

events = pd.read_csv(inicio + 'final_ab_events_upd_us.csv', dtype = {'event_name':'category'}, parse_dates=['event_dt'])

users = pd.read_csv(inicio + 'final_ab_new_users_upd_us.csv', dtype = {'region':'category', 'device':'category'}, parse_dates=['first_date'])

participants = pd.read_csv(inicio + 'final_ab_participants_upd_us.csv', dtype = {'group':'category', 'ab_test':'category'})


### 1.1.3. Informações gerais dos dados:

In [11]:
# Informações gerais:

print('__________________________ Marketing ______________________________\n')
print(marketing.info(memory_usage='deep'))

print('\n__________________________ Events ______________________________\n')
print(events.info(memory_usage='deep'))

print('\n__________________________ Users ______________________________\n')
print(users.info(memory_usage='deep'))

print('\n__________________________ Participants ______________________________\n')
print(participants.info(memory_usage='deep'))

__________________________ Marketing ______________________________

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14 entries, 0 to 13
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype         
---  ------     --------------  -----         
 0   name       14 non-null     object        
 1   regions    14 non-null     category      
 2   start_dt   14 non-null     datetime64[ns]
 3   finish_dt  14 non-null     datetime64[ns]
dtypes: category(1), datetime64[ns](2), object(1)
memory usage: 1.9 KB
None

__________________________ Events ______________________________

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 423761 entries, 0 to 423760
Data columns (total 4 columns):
 #   Column      Non-Null Count   Dtype         
---  ------      --------------   -----         
 0   user_id     423761 non-null  object        
 1   event_dt    423761 non-null  datetime64[ns]
 2   event_name  423761 non-null  category      
 3   details     60314 non-null   float64       


## 1.2. Tratando os valores ausentes

É interessante que a coluna `details` do dataframe `events`só tenha valores em purchase. Vamos verificar isso:

In [12]:
# Verificando os valores ausentes em details para o evento purchase
lines_purchase = events[events['event_name'] == 'purchase']
lines_purchase['details'].isna().sum()

np.int64(0)

Ótimas notícias: Nenhum valor ausente em `details` para o evento purchase

In [13]:
# Verificando os valores NÃO ausentes em details para os outros eventos
lines_outhers = events[~(events['event_name'] == 'purchase')]
lines_outhers.shape[0] - lines_outhers['details'].isna().sum()

np.int64(0)

O que já é bom ainda dar pra melhorar: Todos os demais eventos registram valores ausentes para details. Isso mostra que os dados estão corretos

## 1.3. Tratando os valores duplicados

- Dataframe `marketing`:

In [14]:
expliscitos = marketing.duplicated().sum()

print(f'{expliscitos} duplicados expliscitos')

0 duplicados expliscitos


- Dataframe `events`:

In [15]:
expliscitos = events.duplicated().sum()

print(f'{expliscitos} duplicados expliscitos')

0 duplicados expliscitos


In [16]:
# Procurando bugs do tipo: O mesmo usuário realiza o mesmo tipo de evento no exato mesmo momento.
bugs_duplicated = events[['user_id','event_dt','event_name']].duplicated().sum()

print(f'O mesmo usuário realiza o mesmo tipo de evento no exato mesmo momento: {bugs_duplicated} vezes')


O mesmo usuário realiza o mesmo tipo de evento no exato mesmo momento: 0 vezes


Isso significa que não há esse tipo de problema nos dados

- Dataframe `users`:

In [17]:
expliscitos = users.duplicated().sum()

print(f'{expliscitos} duplicados expliscitos')

0 duplicados expliscitos


In [18]:
# Analisando problemas de usuarios duplicados
clones = users['user_id'].duplicated().sum()

print(f'Existem {clones} usuários duplicados.')

Existem 0 usuários duplicados.


- Dataframe `participants`:

In [19]:
expliscitos = participants.duplicated().sum()

print(f'{expliscitos} duplicados expliscitos')

0 duplicados expliscitos


**Verificando problemas de duplicados impliscitos na coluna `participants`:**

In [20]:
# Verificando problemas do tipo: Um mesmo usuário aparece em mais de um grupo (A e B) dentro do mesmo teste (system e interface):

system = participants[participants['ab_test'] == 'recommender_system_test']
interface = participants[participants['ab_test'] == 'interface_eu_test']

system_problem = system.groupby('user_id')['group'].nunique()
interface_problem = interface.groupby('user_id')['group'].nunique()

print(f'O maior número de aparição de um usuário do teste interface_eu_test em grupos distintos é: {interface_problem.max()}')

print(f'O maior número de aparição de um usuário do recommender_system_test em grupos distintos é: {system_problem.max()}')


O maior número de aparição de um usuário do teste interface_eu_test em grupos distintos é: 1
O maior número de aparição de um usuário do recommender_system_test em grupos distintos é: 1


Portanto, não temos esse tipo de problema

In [21]:
# Verificando problemas do tipo: Um mesmo usuário aparece em mais de um teste diferente.

aparicao = participants.groupby('user_id')['ab_test'].nunique().reset_index()

maximo_aparicoes = aparicao['ab_test'].max()

print(f'O maior número de aparição de um usuário em testes distintos é: {maximo_aparicoes}')

O maior número de aparição de um usuário em testes distintos é: 2


Então existem usuarios em ambos os testes

Agora vamos pensar: por que isso é um problema para o teste A/B?

Imagina que um usuário está no grupo A do teste 1 e no grupo B do teste 2 ao mesmo tempo. O comportamento dele pode ser influenciado pelos dois testes simultaneamente, o que torna impossível saber qual teste causou qual efeito. Diante disso, vamos remover esses usuarios problemáticos dos nossos dataframes `participants, events e users`:

In [22]:
usuarios_problematicos = (aparicao[aparicao['ab_test'] > 1]['user_id']).unique()
total_users = participants['user_id'].nunique()

print(f'{round(100 * len(usuarios_problematicos)/total_users)}% dos usuários são problemáticos, devemos remove-los.')

7% dos usuários são problemáticos, devemos remove-los.


In [23]:
# Removendo os usuarios problematicos:

participants = participants[~(participants['user_id'].isin(usuarios_problematicos))]
events =  events[~(events['user_id'].isin(usuarios_problematicos))]
users =  users[~(users['user_id'].isin(usuarios_problematicos))]


# 2. Realizando a análise exploratória de dados (EDA):

## 2.1. Estudando a conversão em diferentes etapas do funil.

Vamos estudar a conversão para cada grupo do teste recommender_system_test, e depois compara-los:

In [24]:
participants_test = participants[participants['ab_test'] == 'recommender_system_test']

In [25]:
to_funil = events.merge(participants_test, on='user_id')
to_funil.head()

,user_id,event_dt,event_name,details,group,ab_test
0,831887FE7F2D6CBA,2020-12-07 06:50:29,purchase,4.99,A,recommender_system_test
1,3C5DD0288AC4FE23,2020-12-07 19:42:40,purchase,4.99,A,recommender_system_test
2,49EA242586C87836,2020-12-07 06:31:24,purchase,99.99,B,recommender_system_test
3,A640F31CAC7823A6,2020-12-07 18:48:26,purchase,4.99,B,recommender_system_test
4,A9908F62C41613A8,2020-12-07 11:26:47,purchase,9.99,B,recommender_system_test


- Grupo A:

In [26]:
funil_A = (to_funil[to_funil['group'] == 'A']).groupby('event_name')['user_id'].nunique().sort_values(ascending=False).reset_index()
funil_A

C:\Users\Z4r4k1\AppData\Local\Temp\ipykernel_5440\4166450454.py:1: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  funil_A = (to_funil[to_funil['group'] == 'A']).groupby('event_name')['user_id'].nunique().sort_values(ascending=False).reset_index()


,event_name,user_id
0,login,2082
1,product_page,1360
2,purchase,652
3,product_cart,631


Fica claro que o product_cart não é um evento obrigatório do funil, haja vista que há um ganho de tráfego do product_cart para o purchase: Alguns usuários decidiram completar a compra diretamente, sem a necessidade de adição ao carrinho antes. Dessa forma, vamos excluir o evento product_cart do nosso funil (Tanto do Grupo A quanto do B)

In [27]:
# Considerando apenas as etapas obrigatórias do funil:
funil_A = funil_A[funil_A['event_name'] != 'product_cart'].reset_index(drop=True)
funil_A

,event_name,user_id
0,login,2082
1,product_page,1360
2,purchase,652


In [28]:
# Calculando as métricas relativas:
# Usando shift para dividir a linha atual pela anterior
funil_A['conversao'] = (100 * funil_A['user_id'] / funil_A['user_id'].shift(1)).round(2)

# Preenche o primeiro passo (que seria NaN) com 100%
funil_A['conversao'] = funil_A['conversao'].fillna(100)

funil_A

,event_name,user_id,conversao
0,login,2082,100.00
1,product_page,1360,65.32
2,purchase,652,47.94


In [29]:
# Visualisando o Funil A
fig = px.funnel(
    funil_A,
    x="conversao",  
    y="event_name",  
    title="Funil de Conversão - Grupo A",
    labels={"user_id": "Usuários", "event_name": "Etapa"},
    color_discrete_sequence=["#636EFA"], 
)


fig.update_layout(
    title_x=0.5,  
    template="plotly_white",  
)

fig.show()

- Grupo B:

In [30]:
funil_B = (to_funil[to_funil['group'] == 'B']).groupby('event_name')['user_id'].nunique().sort_values(ascending=False).reset_index()
funil_B

C:\Users\Z4r4k1\AppData\Local\Temp\ipykernel_5440\649343626.py:1: FutureWarning:

The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.



,event_name,user_id
0,login,705
1,product_page,397
2,purchase,198
3,product_cart,195


In [31]:
# Considerando apenas as etapas obrigatórias do funil:
funil_B = funil_B[funil_B['event_name'] != 'product_cart'].reset_index(drop=True)
funil_B

,event_name,user_id
0,login,705
1,product_page,397
2,purchase,198


In [32]:
# Calculando as métricas relativas:
# Usando shift para dividir a linha atual pela anterior
funil_B['conversao'] = (100 * funil_B['user_id'] / funil_B['user_id'].shift(1)).round(2)

# Preenche o primeiro passo (que seria NaN) com 100%
funil_B['conversao'] = funil_B['conversao'].fillna(100)

funil_B

,event_name,user_id,conversao
0,login,705,100.00
1,product_page,397,56.31
2,purchase,198,49.87


In [33]:
# Visualisando o Funil B
fig = px.funnel(
    funil_B,
    x="conversao",  
    y="event_name",  
    title="Funil de Conversão - Grupo B",
    labels={"user_id": "Usuários", "event_name": "Etapa"},
    color_discrete_sequence=["#636EFA"], 
)


fig.update_layout(
    title_x=0.5,  
    template="plotly_white",  
)

fig.show()

- Comparando os dois grupos:

In [34]:
funil_A = funil_A.rename(columns = {'user_id':'users_A', 'conversao':'conversao_A'})
funil_B = funil_B.rename(columns = {'user_id':'users_B', 'conversao':'conversao_B'})

funil = funil_A.merge(funil_B, on='event_name')

# Calculando a conversão relativa de B para A:
funil['relative_conversao'] = (100 * (funil['conversao_B'] - funil['conversao_A']) / funil['conversao_A']).round(2)

funil

,event_name,users_A,conversao_A,users_B,conversao_B,relative_conversao
0,login,2082,100.00,705,100.00,0.00
1,product_page,1360,65.32,397,56.31,-13.79
2,purchase,652,47.94,198,49.87,4.03


conclusão: 

O critério de sucesso era um aumento de +10% em cada etapa, e o Grupo B não só ficou longe disso, como ficou abaixo do Grupo A no evento product_page!

## 2.2. O número de eventos por usuário é distribuído igualmente entre as amostras?

In [35]:
# Calculando o numero de eventos por usuario para cada grupo:
events_per_group = (
    to_funil.groupby(["group", "user_id"])["event_name"].count().reset_index()
)

events_A = events_per_group[events_per_group["group"] == "A"]
events_B = events_per_group[events_per_group["group"] == "B"]

# Comparando a distribuição do numero de eventos por usuario de cada grupo:
fig = px.histogram(
    events_per_group,
    x="event_name",
    color="group",
    facet_row="group",
    barmode="overlay",
    labels={
        "event_name": "Número de Eventos por Usuário",
        "group": "Grupo",
        "count": "Contagem de Usuários",
    },
    title="Distribuição do Número de Eventos por Usuário por Grupo",
    height=600,
)

fig.update_traces(opacity=0.75)

color_map = {"A": "#636EFA", "B": "#EF553B"}

for g in events_per_group["group"].unique():
    df_group = events_per_group[events_per_group["group"] == g]
    mean_val = df_group["event_name"].mean()
    median_val = df_group["event_name"].median()
    color = color_map.get(g, "black")

    
    row_idx = 2 if g == "A" else 1

    fig.add_vline(
        x=mean_val,
        line_dash="dash",
        line_color=color,
        annotation_text=f"Média: {mean_val:.2f}",
        annotation_position="top right",
        row=row_idx,
        col=1,
    )

    fig.add_vline(
        x=median_val,
        line_dash="dot",
        line_color=color,
        annotation_text=f"Mediana: {median_val:.2f}",
        annotation_position="top left",
        row=row_idx,
        col=1,
    )

fig.for_each_annotation(
    lambda a: a.update(text=a.text.split("=")[-1]) if "group=" in a.text else None
)

fig.show()

C:\Users\Z4r4k1\AppData\Local\Temp\ipykernel_5440\1697569479.py:3: FutureWarning:

The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.



conclusão:

A análise dos histogramas revela uma discrepância crítica no comportamento e no engajamento dos usuários entre as duas amostras, mostrando que os grupos não estão distribuídos igualmente:

Dominância de Baixo Engajamento no Grupo B: O Grupo B apresenta uma concentração massiva e desproporcional de usuários na faixa mínima de eventos (representada pela primeira barra do gráfico). Essa enorme quantidade de usuários que realizam apenas 1 evento e abandonam a plataforma arrasta a mediana do Grupo B para o valor mínimo (0 ou 1, dependendo do bin) e derruba a média para apenas 1.43 eventos por usuário.

Maior Retenção e Atividade no Grupo A: Em contrapartida, o Grupo A demonstra uma distribuição muito mais saudável e espalhada ao longo da cauda. Os usuários do Grupo A não apenas passam da primeira etapa, como continuam interagindo recorrentemente, o que resulta em uma média de 5.10 eventos por usuário e uma mediana de 4.00.

Em resumo: O Grupo B falhou gravemente em reter os usuários logo após o primeiro contato, enquanto o Grupo A garantiu uma jornada muito mais profunda e engajada. Essa diferença nas métricas centrais indica que o fator que diferencia o Grupo B (seja uma nova feature, página ou campanha) pode estar gerando um impacto negativo na retenção inicial.

## 2.3. Os usuários de ambas as amostras estão presentes?

In [36]:
# Vamos verificar se os usuarios unicos dos grupos A e B de participants do teste recommender_system_test aparecem em events


users_A = participants_test[participants_test['group'] == 'A']['user_id'].unique()
users_B = participants_test[participants_test['group'] == 'B']['user_id'].unique()

users_A = pd.Series(users_A)
users_B = pd.Series(users_B)


users_events = events['user_id'].unique()

isin_A = users_A[users_A.isin(users_events)]
isin_B = users_B[users_B.isin(users_events)]

print(
    f"{round((100 * isin_A.shape[0] / users_A.shape[0]), 2)}% dos usuários do grupo A estão presentes em events"
)
print(
    f"{round((100 * isin_B.shape[0] / users_B.shape[0]), 2)}% dos usuários do grupo B estão presentes em events"
)


100.0% dos usuários do grupo A estão presentes em events
100.0% dos usuários do grupo B estão presentes em events


Concluimos então que os usuarios de ambas as amostras estão presentes

## 2.4. Como o número de eventos é distribuído entre os dias?

In [37]:
events['data'] = events['event_dt'].dt.to_period('D')

events_per_day = events.groupby('data')['event_name'].count().reset_index()


events_per_day["data"] = events_per_day["data"].astype(str)

fig = px.line(
    events_per_day,
    x="data",
    y="event_name",
    labels={"data": "Data", "event_name": "Quantidade de Eventos"},
    title="Distribuição de Eventos por Dia",
    text="event_name",
)

fig.update_traces(textposition="top center")

fig.update_layout(
    yaxis=dict(range=[0, events_per_day["event_name"].max() * 1.15]),
    margin=dict(t=80, b=40, l=40, r=40),
)

fig.show()

conclusão:

Temos um pico entre os dias 14/12 e 21/12, e uma queda drástica após 29/12, devido o término do teste.

## 2.5. Existem quaisquer particularidades nos dados que devemos considerar antes de começar o teste A/B?

**Iremos Verificar se existem os seguintes erros:**

- Campanhas de marketing sobrepostas — Há algum evento de marketing acontecendo durante o período do teste que possa influenciar o comportamento dos usuários?

In [38]:
# Filtrando campanhas que ocorreram no período do teste e afetaram a UE

# Definindo as condições de sobreposição de datas
tempo_sobreposto = (
    ((marketing['start_dt'] <= '2020-12-07') & (marketing['finish_dt'] >= '2021-01-01')) |
    ((marketing['start_dt'] >= '2020-12-07') & (marketing['start_dt'] <= '2021-01-01')) |
    ((marketing['finish_dt'] >= '2020-12-07') & (marketing['finish_dt'] <= '2021-01-01'))
)

# Aplicando os filtros no dataframe
direcionadas_ue = marketing[
    (marketing['regions'].str.contains('EU')) & 
    tempo_sobreposto
]

direcionadas_ue

,name,regions,start_dt,finish_dt
0,Christmas&New Year Promo,"EU, N.America",2020-12-25,2021-01-03


conclusao:

Há uma promoção de ano novo no periodo da campanha

- Tamanho desigual dos grupos — Os grupos A e B têm um número similar de usuários? Uma diferença muito grande pode enviesar os resultados.

In [39]:
funil['proporcao_A'] = (100 * funil['users_A'] / (funil['users_A'] + funil['users_B'])).round(2)
funil['proporcao_B'] = (100 * funil['users_B'] / (funil['users_A'] + funil['users_B'])).round(2)

funil[['event_name', 'proporcao_A', 'proporcao_B']]

,event_name,proporcao_A,proporcao_B
0,login,74.70,25.30
1,product_page,77.40,22.60
2,purchase,76.71,23.29


conclusao:

Em cada evento do Funil, a proporção de usuários do grupo A em relação à quantidade total de usuários dos dois grupos que completaram cada evento gira em torno de 75%, enquanto a proporção de usuários do grupo B gira em torno de 24%. Isso mostra que há uma grande diferença entre o número de usuarios dos grupos A e B.

- Período do teste — O teste ocorre em um período "normal" ou em uma época especial (feriados, promoções sazonais)?

Já vimos que o teste ocorre em um periodo sazonal de dezembro (contemplando promoções de final de ano)

- Número de participantes — O número real de participantes está próximo do esperado?

In [40]:
participants_test['user_id'].nunique()

2788

conclusão:

O número esperado de participantes no teste era de 6000 usuários únicos. Porém, temos apenas 2788

# 3. Avaliando os resultados do teste A/B:

Vamos executar o teste para proporções de cada etapa do funil (vale lembrar que `product_cart` não é uma etapa obrigatória do funil, então não a consideraremos), usando a função `teste_z`:

Para todos os eventos, teremos as seguintes hipóteses:

- ***Hipótese nula***: A conversão entre os grupos foi IGUAL
- ***Hipótese alternativa***: A conversão entre os grupos foi DIFERENTE

In [41]:
funil

,event_name,users_A,conversao_A,users_B,conversao_B,relative_conversao,proporcao_A,proporcao_B
0,login,2082,100.00,705,100.00,0.00,74.70,25.30
1,product_page,1360,65.32,397,56.31,-13.79,77.40,22.60
2,purchase,652,47.94,198,49.87,4.03,76.71,23.29


In [42]:
total_A = participants_test[participants_test['group'] == 'A']['user_id'].nunique() # Total de usuários únicos do grupo A 
total_B = participants_test[participants_test['group'] == 'B']['user_id'].nunique() # Total de usuários únicos do grupo B

- Login

In [43]:
login_A = funil.loc[0,'users_A'] # usuarios unicos do grupo A que realizaram o evento login
login_B = funil.loc[0,'users_B'] # usuarios unicos do grupo B que realizaram o evento login

In [44]:
teste_z(login_A, login_B, total_A, total_B) # Executando o teste z para alpha = 1%

Z-statistic: 1.7176
P-value: 0.0859
Alpha: 0.01


'✅Aceitamos a hipótese nula: A conversão entre os grupos foi IGUAL, logo o teste não trouxe melhorias.'

conclusão:

O teste não trouxe melhorias nas métricas de conversão para o evento `login`

- Product Page		

In [45]:
product_page_A = funil.loc[1,'users_A'] # usuarios unicos do grupo A que realizaram o evento product_page
product_page_B = funil.loc[1,'users_B'] # usuarios unicos do grupo B que realizaram o evento product_page

In [46]:
teste_z(product_page_A, product_page_B, total_A, total_B) # Executando o teste z para alpha = 1%

Z-statistic: 4.3233
P-value: 0.0000
Alpha: 0.01


'❌Rejeitamos a hipótese nula: A conversão entre os grupos foi DIFERENTE'

conclusão: 

Para esse evento, temos as conversões entre os grupos diferentes: O que nos leva a concluir que a conversão do grupo B é 14% menor que a conversão do grupo A para o evento `product_page`, como mostra a coluna `relative_conversao` do dataframe `funil`. Logo, não houve melhoria na conversão do teste para o evento `product_page`.

- Purchase		

In [47]:
purchase_A = funil.loc[2,'users_A'] # usuarios unicos do grupo A que realizaram o evento purchase
purchase_B = funil.loc[2,'users_B'] # usuarios unicos do grupo B que realizaram o evento purchase

In [48]:
teste_z(purchase_A, purchase_B, total_A, total_B) # Executando o teste z para alpha = 1%

Z-statistic: 1.6313
P-value: 0.1028
Alpha: 0.01


'✅Aceitamos a hipótese nula: A conversão entre os grupos foi IGUAL, logo o teste não trouxe melhorias.'

conclusão:

O teste não trouxe melhorias nas métricas de conversão para o evento `purchase`

## conclusão final do teste:

Para preservar capital e não cair em armadilhas estatisticas bem prováveis de acontecer aqui, recomendo a não continuidade desse teste, haja vista que detectamos os seguintes problemas:

- 🎄 Campanha de marketing sobreposta (Christmas & New Year Promo na EU): Período sazonal (dezembro/janeiro — comportamento atípico dos usuários)

  Se há uma campanha de marketing ativa durante o teste, não podemos saber se os resultados são causados pela mudança testada ou pela campanha.

- ⚖️ Grupos muito desiguais (~75% no grupo A vs ~25% no grupo B)

  Se os grupos estão muito desiguais em tamanho, as comparações entre eles são justas, gerando mais certeza para um grupo em comparação com o outro.

- 👥 Poucos participantes (2788 vs 6000 esperados — menos da metade!)

  Se o número de participantes está abaixo da metade do esperado, não conseguimos detectar diferenças reais com confiança.

- Além disso, o nosso teste de proporções mostrou que as modificações não trouxeram nenhuma melhoria para as métricas de conversão dos eventos do funil
